### PR TITLE
SRCH-4915 Implement click tracking in web results

### PR DESCRIPTION
**Pull request by lorawoodford**
## Summary
- Extends click tracking functionality to `web`, `i14y`, and `blended` verticals
 
### Checklist
Please ensure you have addressed all concerns below before marking a PR "ready for review" or before requesting a re-review. If you cannot complete an item below, replace the checkbox with the ⚠️ `:warning:` emoji and explain why the step was not completed.
 
#### Functionality Checks
 
- [x] You have merged the latest changes from the target branch (usually `main`) into your branch.
  
- [x] Your primary commit message is of the format **SRCH-#### \<description\>** matching the associated Jira ticket.

- [x] PR title is either of the format **SRCH-#### \<description\>** matching the associated Jira ticket (i.e. "SRCH-123 implement feature X"), or **Release - SRCH-####, SRCH-####, SRCH-####** matching the Jira ticket numbers in the release.
 
- [x] Automated checks pass. If Code Climate checks do not pass, explain reason for failures:
 
#### Process Checks

- [x] You have specified at least one "Reviewer".

### CODE REVIEW

File: app/javascript/test/ResultGrid.test.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import '@testing-library/jest-dom';
```
Comment (yogesh27) [inline app/javascript/test/ResultGrid.test.tsx:1]: Use below at top of this file to enable the mocking of fetch.
```
import { enableFetchMocks } from 'jest-fetch-mock';
enableFetchMocks();
```
You will need to install it by running `yarn add jest-fetch-mock --dev`

File: app/javascript/test/ResultGrid.test.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    render(<ResultGrid result={result} affiliate='bing_affiliate' query='query' position={1} vertical='web' />);
+    const [firstResult] = Array.from(document.getElementsByClassName('result-title-link'));
+    const clickDataJson = JSON.parse(firstResult.getAttribute('data-click') || '{}');
+
+    expect(clickDataJson.affiliate).toBe('bing_affiliate');
+    expect(clickDataJson.url).toBe('https://www.search.gov');
+    expect(clickDataJson.module_code).toBe('BWEB');
+    expect(clickDataJson.position).toBe(1);
+    expect(clickDataJson.query).toBe('query');
+    expect(clickDataJson.vertical).toBe('web');
```
Comment (yogesh27) [inline app/javascript/test/ResultGrid.test.tsx:29]:
```
const title = screen.getByText(/test result 1/i);
fireEvent.click(title);
```

File: app/javascript/test/ResultGrid.test.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    render(<ResultGrid result={result} affiliate='searchgov_affiliate' query='query' position={2} vertical='i14y' />);
+    const [firstResult] = Array.from(document.getElementsByClassName('result-title-link'));
+    const clickDataJson = JSON.parse(firstResult.getAttribute('data-click') || '{}');
+
+    expect(clickDataJson.affiliate).toBe('searchgov_affiliate');
+    expect(clickDataJson.url).toBe('https://www.search.gov');
+    expect(clickDataJson.module_code).toBe('I14Y');
+    expect(clickDataJson.position).toBe(2);
+    expect(clickDataJson.query).toBe('query');
+    expect(clickDataJson.vertical).toBe('i14y');
```
Comment (yogesh27) [inline app/javascript/test/ResultGrid.test.tsx:42]:
```
const title = screen.getByText(/test result 1/i);
fireEvent.click(title);
```

File: app/javascript/test/ResultGrid.test.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    render(<ResultGrid result={blendedresult} affiliate='blended_affiliate' query='query' position={3} vertical='blended' />);
+    const [firstResult] = Array.from(document.getElementsByClassName('result-title-link'));
+    const clickDataJson = JSON.parse(firstResult.getAttribute('data-click') || '{}');
+
+    expect(clickDataJson.affiliate).toBe('blended_affiliate');
+    expect(clickDataJson.url).toBe('https://www.search.gov');
+    expect(clickDataJson.module_code).toBe('AIDOC');
+    expect(clickDataJson.position).toBe(3);
+    expect(clickDataJson.query).toBe('query');
+    expect(clickDataJson.vertical).toBe('blended');
```
Comment (yogesh27) [inline app/javascript/test/ResultGrid.test.tsx:55]:
```
const title = screen.getByText(/test result 1/i);
fireEvent.click(title);
```

File: .rubocop.yml [hunk 5]
```diff
@@ -8,14 +8,6 @@ inherit_mode:
     - Exclude
     - AllowedMethods
 
-Layout/AlignHash:
```
Comment (lorawoodford) [inline .rubocop.yml:11]: `AlignHash` was renamed to `HashAlignment` back in 2019.  We don't need both: https://github.com/rubocop/rubocop/commit/7ab80f9e1d262ff34fe4a1ea1a752a22bec26310

File: app/controllers/clicked_controller.rb [hunk 6]
```diff
@@ -23,6 +23,7 @@ def click_class
   end
 
   def permitted_params
+    params.delete(:clicked)
```
Comment (lorawoodford) [inline app/controllers/clicked_controller.rb:26]: See my question in SRCH-4881.  In a perfect world I'd like not to need this, but I can't figure out a workaround.  Input welcome.

Comment (yogesh27) [reply app/controllers/clicked_controller.rb:26]: Hi @stevenbarragan @luismetzger @krbhavith - any thoughts on this?

File: spec/models/results_with_body_and_description_post_processor_spec.rb
Comment (yogesh27) [inline spec/models/results_with_body_and_description_post_processor_spec.rb:1]: @stevenbarragan can you review this piece of code?

**Review by yogesh27 — COMMENTED** [review]
Hi @lorawoodford - Javascript related changes looks good to me. There is some piece of code which I think need to be reviewed by other members, so I have tagged them.

File: app/controllers/clicked_controller.rb (hunk 6 above)
Comment (stevenbarragan) [reply app/controllers/clicked_controller.rb:26]: It's hard to follow where this is coming from. From the logs from [here](https://cm-jira.usa.gov/browse/SRCH-4881?focusedId=1850760&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1850760)
```
16:57:07 web.1  |   Parameters: {"affiliate"=>"test_affiliate", "url"=>"https://search.gov", "module_code"=>"BOOS", "position"=>1, "query"=>"best bet", "vertical"=>"web", "clicked"=>{"affiliate"=>"test_affiliate", "url"=>"https://search.gov", "module_code"=>"BOOS", "position"=>1, "query"=>"best bet", "vertical"=>"web"}}
16:57:07 web.1  | Unpermitted parameter: :clicked. Context: { controller: ClickedController, action: create, request: #<ActionDispatch::Request:0x00007f89c83cc950>, params: {"affiliate"=>"test_affiliate", "url"=>"https://search.gov", "module_code"=>"BOOS", "position"=>1, "query"=>"best bet", "vertical"=>"web", "controller"=>"clicked", "action"=>"create", "clicked"=>{"affiliate"=>"test_affiliate", "url"=>"https://search.gov", "module_code"=>"BOOS", "position"=>1, "query"=>"best bet", "vertical"=>"web"}} }
```

I can only tell that the params are been duplicated on the `clicked` key. Are the logs from [this request](https://github.com/GSA/search-gov/pull/1488/files#diff-6627eb1616884db52ea1b6eeac2a9cd658aceec88de7cfaeef31a97223ba0403R81)? Where does `params["clicked"]` gets populated?

**Review by stevenbarragan — CHANGES_REQUESTED** [review]

File: spec/models/results_with_body_and_description_post_processor_spec.rb [hunk 7]
```diff
@@ -68,5 +68,37 @@
         end
       end
     end
+
+    context 'when there are news item results' do
+      subject(:normalized_results) { described_class.new(results, _val: nil).normalized_results(5) }
+
+      let(:results) do
+        results = []
+        5.times { |index| results << NewsItem.new(title: "title #{index}", description: "content #{index}", link: "http://foo.gov/#{index}") }
```
[the remark this answers is not in the document]
Comment (lorawoodford) [reply spec/models/results_with_body_and_description_post_processor_spec.rb]: ✅

File: spec/models/results_with_body_and_description_post_processor_spec.rb [hunk 8]
```diff
@@ -68,5 +68,37 @@
...
+        end
+      end
+    end
+
+    context 'when there are indexed document results' do
+      subject(:normalized_results) { described_class.new(results, _val: nil).normalized_results(5) }
+
+      let(:results) do
+        results = []
+        5.times { |index| results << IndexedDocument.new(title: "title #{index}", description: "content #{index}", url: "http://foo.gov/#{index}") }
```
[the remark this answers is not in the document]
Comment (lorawoodford) [reply spec/models/results_with_body_and_description_post_processor_spec.rb]: ✅

File: app/javascript/components/Results/ResultGrid/ResultGrid.tsx [hunk 9]
```diff
@@ -55,7 +68,10 @@ export const ResultGrid = ({ result }: ResultProps) => {
           {result.publishedAt   && <span className='published-date'>{result.publishedAt}</span>}
           <div className='result-title'>
             <h2 className='result-title-label'>
-              <a href={result.url} className='result-title-link'>
+              <a href={result.url}
+                className='result-title-link'
+                data-click={clickData(affiliate, module, query, position, result.url, vertical)}
+                onClick={(event) => clickTracking((event.target as HTMLLinkElement).getAttribute('data-click') || '{}')}>
```
[the remark this answers is not in the document]
Comment (lorawoodford) [reply app/javascript/components/Results/ResultGrid/ResultGrid.tsx:74]: That's how I had it in my initial implementation, but changed it here for a few reasons:

1. I found it handy to be able to inspect the clickdata in the page itself.  And I *suspect* it might be handy for the admin team as well as they _may_ be used to being able to do so as the current SERP implementation sets the click data via a data attribute.  It might be worthwhile for me to check that assumption.
2. It made testing the module assignment for blended results done [here](https://github.com/GSA/search-gov/pull/1488/files#diff-1cb55c059efac4ba4ed9b9695bbb241dee5a3885a9ff1af03bb0b47f66189f96R49-R55) much simpler: https://github.com/GSA/search-gov/pull/1488/files#diff-23b312b55c739e2aff9ccf80cad2b6918c634292cffafd97fffd97461aef345bR26-R31  That set, dynamic setting of module code is ONLY relevant within this module and only in the blended vertical.  This could likely be tested within the onclick itself, but would probably be beyond my capacity.

I think question 1 is to figure out whether having that visible data attribute is of use to anyone, so I'll ask.

[the remark this answers is not in the document]
Comment (lorawoodford) [reply app/javascript/components/Results/ResultGrid/ResultGrid.tsx]: Alright, @stevenbarragan , @yogesh27 I've done what Steven suggested and passed the data directly to `clickTracking`.  This works, and codecoverage is happy, but it feels wrong to have tests entirely devoid of expectations here: https://github.com/GSA/search-gov/blob/3b123a22a4c8969f36aed869d30589bf996909e1/app/javascript/test/ResultGrid.test.tsx

Does this seem alright?

**Comment by lorawoodford** [comment]
@stevenbarragan This is ready for another look.  Re this:

https://github.com/GSA/search-gov/pull/1488#discussion_r1509537889

After some digging I learned something new.  This is being populated via default Rails magic for [parameter wrapping](https://api.rubyonrails.org/classes/ActionController/ParamsWrapper.html).  Disabling it in this controller resolves the need to delete the duplicated, unpermitted `clicked` param.  I also attempted to go down the path of updating `pemitted_params` to the more standard:

```
  def permitted_params
    params.require(:clicked).permit(
      :url,
      :query,
      :position,
      :module_code,
      :affiliate,
      :vertical
    )
  end
```

but this broke tests all over for both API and legacy SERP click tracking and felt like a riskier move than we'd want to make right now.  We may want to revisit in the future once the legacy SERP is removed.

So, assuming the above sounds alright, the only remaining question I have for you (and @yogesh27 ) is regarding the adequacy of this test coverage: https://github.com/GSA/search-gov/pull/1488#discussion_r1509640737

File: app/javascript/components/Results/ResultGrid/ResultGrid.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (yogesh27) [reply app/javascript/components/Results/ResultGrid/ResultGrid.tsx]: Hi @lorawoodford, good point you raised - I think what we could do is to check if the clickTracking method was called when the anchor tag is clicked like:
```
it("should call clickTracking when link is clicked", () => {
  const myClickTracking = jest.fn();
  const { getByText } = render(
    <YourComponent clickTracking={myClickTracking} />
  );
  const link = getByText("test result 1");
  fireEvent.click(link);
  expect(clickTracking).toHaveBeenCalledTimes(1);
});
```
but this would need a major code change across all SERP to refractor the code for this, and also similar test would be needed while implementing clickTracking for mobile view as we will be making the whole result wrapper to be clickable. I think this implementation can be in its own JIRA ticket.

@stevenbarragan your thoughts on this?

File: app/javascript/utils/index.ts [hunk 10]
```diff
@@ -73,7 +73,8 @@ export const clickTracking = (affiliate: string, module: string, query: string,
       module_code: module,
       position,
       query,
-      vertical
+      vertical,
+      type: 'click'
```
Comment (lorawoodford) [inline app/javascript/utils/index.ts:77]: While setting up AT I discovered another undocumented requirement based on current functionality.  All current SERP click tracking has a "type" of 'click' (as opposed to 'search' for search logging).

File: app/javascript/components/Results/ResultGrid/ResultGrid.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (stevenbarragan) [reply app/javascript/components/Results/ResultGrid/ResultGrid.tsx]: You're right @lorawoodford. This tests are not testing!

Can we somehow do something like?

```js
fireEvent.click(title);
expect(fetch).toHaveBeenCalledWith(...);
expect(fetch).toHaveBeenCalledTimes(1);
```

I like your suggestion @yogesh27 as well.

[the remark this answers is not in the document]
Comment (stevenbarragan) [reply app/javascript/components/Results/ResultGrid/ResultGrid.tsx]: I wasn't very clear. I'm suggesting we can use the same test and validate the call to fetch. e.g.

```js
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve({...}),
    })
  );

  beforeEach(() => {
    fetch.mockClear();
  });

  it('sets correct Bing click data', () => {
    render(<ResultGrid result={result} affiliate='bing_affiliate' query='query' position={1} vertical='web' />);

    const title = screen.getByText(/test result 1/i);
    fireEvent.click(title);

    expect(fetch).toHaveBeenCalledWith(...);
    expect(fetch).toHaveBeenCalledTimes(1);
  });
```

[the remark this answers is not in the document]
Comment (lorawoodford) [reply app/javascript/components/Results/ResultGrid/ResultGrid.tsx]: @stevenbarragan Alright, I think that got me where I needed to be.  Tests updated in latest commit.

**Review by stevenbarragan — APPROVED** [review]
LGTM 🙏🏽
